### PR TITLE
Flush commit metadata even when punning fails

### PR DESCRIPTION
**Pull request by p0mvn**
## Description

Closes: #XXXX

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Comment by UnityChaos** [comment]
Looks reasonable to me, but @ValarDragon may object because this might cause other issues that I'm not aware of. Will rerun the "crash when trying to prune snapshot height" test with the branch tho.

File: store/rootmulti/store.go [hunk 1]
```diff
@@ -382,6 +382,14 @@ func (rs *Store) Commit() types.CommitID {
 
 	rs.lastCommitInfo = commitStores(version, rs.stores)
 
+	var err error
```
Comment (UnityChaos) [inline store/rootmulti/store.go]:
```suggestion
	var pruneErr error
```

File: store/rootmulti/store.go [hunk 2]
```diff
@@ -398,11 +406,9 @@ func (rs *Store) Commit() types.CommitID {
 
 	// batch prune if the current height is a pruning interval height
 	if rs.pruningOpts.Interval > 0 && version%int64(rs.pruningOpts.Interval) == 0 {
-		rs.pruneStores()
+		err = rs.pruneStores()
```
Comment (UnityChaos) [inline store/rootmulti/store.go]:
```suggestion
		pruneErr = rs.pruneStores()
```

File: store/rootmulti/store.go [hunk 3]
```diff
@@ -382,6 +382,14 @@ func (rs *Store) Commit() types.CommitID {
 
 	rs.lastCommitInfo = commitStores(version, rs.stores)
 
+	var err error
+	defer func ()  {
+		flushMetadata(rs.db, version, rs.lastCommitInfo, rs.pruneHeights)
+		if err != nil {
+			panic(err)
```
Comment (UnityChaos) [inline store/rootmulti/store.go]:
```suggestion
		if pruneErr != nil {
			panic(pruneErr)
```

**Review by UnityChaos — APPROVED** [review]
Minor rename based on suggestion from @ValarDragon.
Other than that this LGTM :tada: 

----

Tested this on mainnet with settings:
```
pruning = "custom"

# These are applied if and only if the pruning strategy is custom.
# pruning-keep-recent = N means keep all of the last N states
pruning-keep-recent = "100"
# pruning-keep-every = N means keep every Nth state, in addition to keep-recent
pruning-keep-every = "0"
# pruning-interval = N means we delete old states from disk every Nth block.
pruning-interval = "10"
```
and
```
###############################################################################
###                        State Sync Configuration                         ###
###############################################################################

# State sync snapshots allow other nodes to rapidly join the network without replaying historical
# blocks, instead downloading and applying a snapshot of the application state at a given height.
[state-sync]

# snapshot-interval specifies the block interval at which local state sync snapshots are
# taken (0 to disable). Must be a multiple of pruning-keep-every.
snapshot-interval = 30

# snapshot-keep-recent specifies the number of recent snapshots to keep and serve (0 to keep all).
snapshot-keep-recent = 2
```

This consistently causes a failure on pruning, and I was able to confirm that flushing the metadata with this patch is successfully fixing the halt on replay we were finding earlier.

The settings I'm using here (keep recent = 100, snapshot interval 30, keep 2 snapshots) shouldn't be causing crashes actually, so that looks to be another bug that I will investigate separately.

**Comment by p0mvn** [comment]
That's great to hear. Thanks for reviewing and testing.

Since we know that this change fixes the earlier problem related to 6.3.0 only, now we can try addressing the "active reader problem" related to both 6.2.0 and 6.3.0. I will draft it up tomorrow morning. I think it's relate to not having synchronization in IAVL's nodedb

**Review by ValarDragon — COMMENTED** [review]
LGTM! Nice catch

**Pull request merged**